bootstrap/internal/controllers/rke2config_controller.go

```diff
@@ -278,6 +278,14 @@ type Scope struct {
 	ControlPlane         *controlplanev1.RKE2ControlPlane
 }
 
+func (s *Scope) getDesiredVersion() string {
+	if s.Machine.Spec.Version != nil && bsutil.IsRKE2Version(*s.Machine.Spec.Version) {
```

Comment:
If there is a value `s.Machine.Spec.Version` but it doesn't have the rke2 suffix then it will be ignored? Should we raise an error in this situation?

Reply:
I think we should get rid of the IsRKE2Version and add a validation on the CRD itself. It will simplify things

Reply:
@richardcase Currently we create machines with plain k8s version by removing RKE2 suffix. This function ensures nothing breaks existing clusters, we need to read version from agentConfig. After users migrate to `spec.Version` field and rollout new machines, `machine.Spec.Version` will always be RKE2 version.

Reply:
@Danil-Grigorev we can't add CRD validation for core CAPI resources, the version for workers will come from MD. It's possible to create a custom webhook but I don't know if want to go this route.

Reply:
We can still add a validation to the RKE2 Control Plane object, I will update the PR

Reply:
CP provider doesn't care what version workers have, these will not appear here

Reply:
We pass value from machine.Spec.Version to the bootstrap script for all type of machines, it means that this value must be a valid RKE2 version

Reply:
I added the validation in the CRD but we still need to keep checking of RKE2 version in these functions for compatibility with already created clusters.